flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogDelegateStateTest.java

```diff
@@ -0,0 +1,134 @@
+    private MockEnvironment env;
+
+    @Parameterized.Parameters
+    public static List<Supplier<AbstractStateBackend>> delegatedStateBackend() {
+        return Arrays.asList(HashMapStateBackend::new, EmbeddedRocksDBStateBackend::new);
+    }
+
+    @Parameterized.Parameter public Supplier<AbstractStateBackend> backend;
+
+    @Before
+    public void before() {
+        env = MockEnvironment.builder().build();
+    }
+
+    @After
+    public void after() {
+        IOUtils.closeQuietly(env);
+    }
+
+    @Test
+    public void testDelegatingValueState() throws Exception {
+        testDelegatingState(
+                new ValueStateDescriptor<>("id", String.class), ChangelogValueState.class);
+    }
+
+    @Test
+    public void testDelegatingListState() throws Exception {
+        testDelegatingState(
+                new ListStateDescriptor<>("id", String.class), ChangelogListState.class);
+    }
+
+    @Test
+    public void testDelegatingMapState() throws Exception {
+        testDelegatingState(
+                new MapStateDescriptor<>("id", Integer.class, String.class),
+                ChangelogMapState.class);
+    }
+
+    @Test
+    public void testDelegatingReducingState() throws Exception {
+        testDelegatingState(
+                new ReducingStateDescriptor<>(
+                        "id", (value1, value2) -> value1 + "," + value2, String.class),
+                ChangelogReducingState.class);
+    }
+
+    @Test
+    public void testDelegatingAggregatingState() throws Exception {
+        testDelegatingState(
+                new AggregatingStateDescriptor<>(
+                        "my-state",
+                        new StateBackendTestBase.MutableAggregatingAddingFunction(),
+                        StateBackendTestBase.MutableLong.class),
+                ChangelogAggregatingState.class);
+    }
+
+    private void testDelegatingState(StateDescriptor descriptor, Class<?> stateClass)
+            throws Exception {
+        KeyedStateBackend<Integer> delegatedBackend = createKeyedBackend(backend.get(), env);
+        KeyedStateBackend<Integer> changelogBackend =
+                createKeyedBackend(new ChangelogStateBackend(backend.get()), env);
+
+        try {
+            State state =
+                    changelogBackend.getPartitionedState(
+                            VoidNamespace.INSTANCE, VoidNamespaceSerializer.INSTANCE, descriptor);
+
+            assertSame(state.getClass(), stateClass);
+            assertSame(
+                    ((AbstractChangelogState<?, ?, ?, ?>) state).getDelegatedState().getClass(),
+                    delegatedBackend
+                            .getPartitionedState(
+                                    VoidNamespace.INSTANCE,
+                                    VoidNamespaceSerializer.INSTANCE,
+                                    descriptor)
+                            .getClass());
+        } finally {
+            delegatedBackend.dispose();
```

Comment:
We'd better add not null check and then disposing those keyed state backends.

Reply:
OK